File: 0310-minimum-height-trees/0310-minimum-height-trees.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        if (n == 1) return {0}; // If there's only one node, it's the root of the MHT

        vector<int> in(n, 0);
        vector<vector<int>> adjList(n);

        // Build the graph
        for (auto edge : edges) {
            adjList[edge[0]].push_back(edge[1]);
            adjList[edge[1]].push_back(edge[0]);
            in[edge[0]]++;
            in[edge[1]]++;
        }

        // Initialize the first layer of leaves
        queue<int> q;
        for (int i = 0; i < n; i++) {
            if (in[i] == 1) q.push(i);
        }

        int remainingNodes = n;

        // Trim the leaves until reaching the centroids
        while (remainingNodes > 2) {
            int leavesCount = q.size();
            remainingNodes -= leavesCount;
            for (int i = 0; i < leavesCount; i++) {
                int leaf = q.front();
                q.pop();
                for (int neighbor : adjList[leaf]) {
                    in[neighbor]--;
                    if (in[neighbor] == 1) q.push(neighbor);
                }
            }
        }

        // Collect the remaining nodes, which are the centroids of the tree
        vector<int> res;
        while (!q.empty()) {
            res.push_back(q.front());
            q.pop();
        }

        return res;
    }
};

File: 0310-minimum-height-trees/0310-minimum-height-trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        adjList = defaultdict(list)
        for src,dst in edges:
            adjList[src].append(dst)
            adjList[dst].append(src)

        edge_cnt = {}
        leaves = deque()

        for src, nei in adjList.items():
            if len(nei) == 1:
                leaves.append(src)
            edge_cnt[src] = len(nei)
        

        while leaves:
            if (n <= 2):
                return list(leaves)
            n -= len(leaves)
            for i in range(len(leaves)):
                node = leaves.popleft()
                for nei in adjList[node]:
                    edge_cnt[nei] -= 1
                    if edge_cnt[nei] == 1:
                        leaves.append(nei)
        
            


        